syntax = "proto3";
package farotracker;

import "google/protobuf/empty.proto";
import "google/protobuf/wrappers.proto";

option java_multiple_files = true;
option java_package = "me.nicholasnadeau.farotracker";
option java_outer_classname = "FaroTrackerProto";

// define service
service FaroTrackerService {
    rpc MoveCartesian (CartesianPosition) returns (google.protobuf.Empty) {
    }

    rpc MoveCartesianRel (CartesianPosition) returns (google.protobuf.Empty) {
    }

    rpc MoveSpherical (SphericalPosition) returns (google.protobuf.Empty) {
    }

    rpc MoveSphericalRel (SphericalPosition) returns (google.protobuf.Empty) {
    }

    rpc MoveHome (google.protobuf.Empty) returns (google.protobuf.Empty) {
    }

    rpc Search (google.protobuf.DoubleValue) returns (google.protobuf.BoolValue) {
    }

    rpc MeasurePoint (google.protobuf.Empty) returns (Measure) {
    }

    rpc MeasureLevel (google.protobuf.Empty) returns (Level) {
    }

    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty) {
    }

    rpc Initialize (google.protobuf.Empty) returns (google.protobuf.Empty) {
    }

    rpc Stop (google.protobuf.Empty) returns (google.protobuf.Empty) {
    }

    rpc IsTargetDetected (google.protobuf.Empty) returns (google.protobuf.BoolValue) {
    }

    rpc ChangeLedState (LedState) returns (google.protobuf.Empty) {
    }
}

message CartesianPosition {
    double x = 1;
    double y = 2;
    double z = 3;
}

message SphericalPosition {
    double azimuth = 1;
    double zenith = 2;
    double distance = 3;
}

message Measure {
    CartesianPosition position = 1;
    double rms = 2;
    double temperature = 3;
    bool is_success = 4;
}

message Level {
    double Rx = 1;
    double Ry = 2;
    double Rz = 3;
    bool is_success = 4;
}

message LedState {
    Led led = 1;
    LedBlink ledBlink = 2;
}

enum Led {
    ORANGE = 0;
    BLUE = 1;
    BOTH = 2;
}

enum LedBlink {
    OFF = 0;
    ON = 1;
    SLOW = 2;
    NORMAL = 3;
    FAST = 4;
}
